File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/UtilsTest.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


package com.starrocks.sql.optimizer;


import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.starrocks.analysis.JoinOperator;
import com.starrocks.catalog.Column;
import com.starrocks.catalog.OlapTable;
import com.starrocks.catalog.Partition;
import com.starrocks.catalog.Type;
import com.starrocks.common.DdlException;
import com.starrocks.common.FeConstants;
import com.starrocks.qe.ConnectContext;
import com.starrocks.server.GlobalStateMgr;
import com.starrocks.sql.ast.CreateDbStmt;
import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
import com.starrocks.sql.optimizer.operator.logical.LogicalValuesOperator;
import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
import com.starrocks.sql.optimizer.statistics.ColumnStatistic;
import com.starrocks.statistic.StatsConstants;
import com.starrocks.utframe.StarRocksAssert;
import com.starrocks.utframe.UtFrameUtils;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class UtilsTest {
    private static final String DEFAULT_CREATE_TABLE_TEMPLATE = ""
            + "CREATE TABLE IF NOT EXISTS `table_statistic_v1` (\n"
            + "  `table_id` bigint NOT NULL,\n"
            + "  `column_name` varchar(65530) NOT NULL,\n"
            + "  `db_id` bigint NOT NULL,\n"
            + "  `table_name` varchar(65530) NOT NULL,\n"
            + "  `db_name` varchar(65530) NOT NULL,\n"
            + "  `row_count` bigint NOT NULL,\n"
            + "  `data_size` bigint NOT NULL,\n"
            + "  `distinct_count` bigint NOT NULL,\n"
            + "  `null_count` bigint NOT NULL,\n"
            + "  `max` varchar(65530) NOT NULL,\n"
            + "  `min` varchar(65530) NOT NULL,\n"
            + "  `update_time` datetime NOT NULL\n"
            + "  )\n"
            + "ENGINE=OLAP\n"
            + "UNIQUE KEY(`table_id`,  `column_name`, `db_id`)\n"
            + "DISTRIBUTED BY HASH(`table_id`, `column_name`, `db_id`) BUCKETS 2\n"
            + "PROPERTIES (\n"
            + "\"replication_num\" = \"1\",\n"
            + "\"in_memory\" = \"false\",\n"
            + "\"storage_format\" = \"V2\"\n"
            + ");";

    private static ConnectContext connectContext;
    private static StarRocksAssert starRocksAssert;

    protected static void setTableStatistics(OlapTable table, long rowCount) {
        for (Partition partition : table.getAllPartitions()) {
            partition.getBaseIndex().setRowCount(rowCount);
        }
    }

    @BeforeClass
    public static void beforeClass() throws Exception {
        FeConstants.default_scheduler_interval_millisecond = 1;
        UtFrameUtils.createMinStarRocksCluster();

        // create connect context
        connectContext = UtFrameUtils.createDefaultCtx();
        starRocksAssert = new StarRocksAssert(connectContext);
        String dbName = "test";
        starRocksAssert.withDatabase(dbName).useDatabase(dbName);

        connectContext.getSessionVariable().setMaxTransformReorderJoins(8);
        connectContext.getSessionVariable().setOptimizerExecuteTimeout(30000);
        connectContext.getSessionVariable().setEnableReplicationJoin(false);

        starRocksAssert.withTable("CREATE TABLE `t0` (\n" +
                "  `v1` bigint NULL COMMENT \"\",\n" +
                "  `v2` bigint NULL COMMENT \"\",\n" +
                "  `v3` bigint NULL\n" +
                ") ENGINE=OLAP\n" +
                "DUPLICATE KEY(`v1`, `v2`, v3)\n" +
                "DISTRIBUTED BY HASH(`v1`) BUCKETS 3\n" +
                "PROPERTIES (\n" +
                "\"replication_num\" = \"1\",\n" +
                "\"in_memory\" = \"false\",\n" +
                "\"storage_format\" = \"DEFAULT\"\n" +
                ");");

        starRocksAssert.withTable("CREATE TABLE `t1` (\n" +
                "  `v1` bigint NULL COMMENT \"\",\n" +
                "  `v2` bigint NULL COMMENT \"\",\n" +
                "  `v3` bigint NULL\n" +
                ") ENGINE=OLAP\n" +
                "AGGREGATE KEY(`v1`, `v2`, v3)\n" +
                "DISTRIBUTED BY HASH(`v1`) BUCKETS 3\n" +
                "PROPERTIES (\n" +
                "\"replication_num\" = \"1\",\n" +
                "\"in_memory\" = \"false\",\n" +
                "\"storage_format\" = \"DEFAULT\"\n" +
                ");");

        CreateDbStmt dbStmt = new CreateDbStmt(false, StatsConstants.STATISTICS_DB_NAME);
        try {
            GlobalStateMgr.getCurrentState().getMetadata().createDb(dbStmt.getFullDbName());
        } catch (DdlException e) {
            return;
        }
        starRocksAssert.useDatabase(StatsConstants.STATISTICS_DB_NAME);
        starRocksAssert.withTable(DEFAULT_CREATE_TABLE_TEMPLATE);
        FeConstants.runningUnitTest = true;
    }

    @Test
    public void extractConjuncts() {
        ScalarOperator root =
                new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                        ConstantOperator.createBoolean(false),
                        new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                                ConstantOperator.createBoolean(true),
                                new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                                        ConstantOperator.createInt(1),
                                        ConstantOperator.createInt(2))));

        List<ScalarOperator> list = Utils.extractConjuncts(root);

        assertEquals(4, list.size());
        assertFalse(((ConstantOperator) list.get(0)).getBoolean());
        assertTrue(((ConstantOperator) list.get(1)).getBoolean());
        assertEquals(1, ((ConstantOperator) list.get(2)).getInt());
        assertEquals(2, ((ConstantOperator) list.get(3)).getInt());
    }

    @Test
    public void extractColumnRef() {
        ScalarOperator root =
                new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                        ConstantOperator.createBoolean(false),
                        new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                                new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.EQ,
                                        new ColumnRefOperator(3, Type.INT, "hello", true),
                                        ConstantOperator.createInt(1)),
                                new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.AND,
                                        new ColumnRefOperator(1, Type.INT, "name", true),
                                        new ColumnRefOperator(2, Type.INT, "age", true))));

        List<ColumnRefOperator> list = Utils.extractColumnRef(root);

        assertEquals(3, list.size());
        assertTrue(list.get(0).isVariable());
        assertTrue(list.get(1).isVariable());
        assertTrue(list.get(2).isVariable());
        assertEquals(3, list.get(0).getId());
        assertEquals(1, list.get(1).getId());
        assertEquals(2, list.get(2).getId());
    }

    @Test
    public void compoundAnd1() {
        ScalarOperator tree1 = Utils.compoundAnd(ConstantOperator.createInt(1),
                ConstantOperator.createInt(2),
                ConstantOperator.createInt(3),
                ConstantOperator.createInt(4),
                ConstantOperator.createInt(5));

        assertEquals(CompoundPredicateOperator.CompoundType.AND, ((CompoundPredicateOperator) tree1).getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) tree1.getChild(0)).getCompoundType());
        assertEquals(5, ((ConstantOperator) tree1.getChild(1)).getInt());

        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) tree1.getChild(0).getChild(0)).getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) tree1.getChild(0).getChild(1)).getCompoundType());

        assertEquals(1, ((ConstantOperator) tree1.getChild(0).getChild(0).getChild(0)).getInt());
        assertEquals(2, ((ConstantOperator) tree1.getChild(0).getChild(0).getChild(1)).getInt());

        assertEquals(3, ((ConstantOperator) tree1.getChild(0).getChild(1).getChild(0)).getInt());
        assertEquals(4, ((ConstantOperator) tree1.getChild(0).getChild(1).getChild(1)).getInt());
    }

    @Test
    public void compoundAnd2() {
        ScalarOperator tree1 = Utils.compoundAnd(ConstantOperator.createInt(1),
                ConstantOperator.createInt(2),
                ConstantOperator.createInt(3),
                ConstantOperator.createInt(4));

        assertEquals(CompoundPredicateOperator.CompoundType.AND, ((CompoundPredicateOperator) tree1).getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) tree1.getChild(0)).getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) tree1.getChild(1)).getCompoundType());

        assertEquals(1, ((ConstantOperator) tree1.getChild(0).getChild(0)).getInt());
        assertEquals(2, ((ConstantOperator) tree1.getChild(0).getChild(1)).getInt());

        assertEquals(3, ((ConstantOperator) tree1.getChild(1).getChild(0)).getInt());
        assertEquals(4, ((ConstantOperator) tree1.getChild(1).getChild(1)).getInt());
    }

    @Test
    public void compoundAnd3() {
        ScalarOperator tree1 = Utils.compoundAnd(ConstantOperator.createInt(1),
                ConstantOperator.createInt(2),
                ConstantOperator.createInt(3),
                ConstantOperator.createInt(4),
                ConstantOperator.createInt(5),
                ConstantOperator.createInt(6));

        assertEquals(CompoundPredicateOperator.CompoundType.AND, ((CompoundPredicateOperator) tree1).getCompoundType());
        CompoundPredicateOperator leftChild = (CompoundPredicateOperator) tree1.getChild(0);
        CompoundPredicateOperator rightChild = (CompoundPredicateOperator) tree1.getChild(1);

        assertEquals(CompoundPredicateOperator.CompoundType.AND, leftChild.getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND, rightChild.getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) leftChild.getChild(0)).getCompoundType());
        assertEquals(CompoundPredicateOperator.CompoundType.AND,
                ((CompoundPredicateOperator) leftChild.getChild(1)).getCompoundType());

        assertEquals(1, ((ConstantOperator) leftChild.getChild(0).getChild(0)).getInt());
        assertEquals(2, ((ConstantOperator) leftChild.getChild(0).getChild(1)).getInt());

        assertEquals(3, ((ConstantOperator) leftChild.getChild(1).getChild(0)).getInt());
        assertEquals(4, ((ConstantOperator) leftChild.getChild(1).getChild(1)).getInt());

        assertEquals(5, ((ConstantOperator) rightChild.getChild(0)).getInt());
        assertEquals(6, ((ConstantOperator) rightChild.getChild(1)).getInt());
    }

    @Test
    public void unknownStats1() {
        GlobalStateMgr globalStateMgr = connectContext.getGlobalStateMgr();

        OlapTable t0 = (OlapTable) globalStateMgr.getDb("test").getTable("t0");
        setTableStatistics(t0, 10);
        GlobalStateMgr.getCurrentStatisticStorage().addColumnStatistic(t0, "v1",
                new ColumnStatistic(1, 1, 0, 1, 1));

        Map<ColumnRefOperator, Column> columnRefMap = new HashMap<>();
        columnRefMap.put(new ColumnRefOperator(1, Type.BIGINT, "v1", true),
                t0.getBaseColumn("v1"));
        columnRefMap.put(new ColumnRefOperator(2, Type.BIGINT, "v2", true),
                t0.getBaseColumn("v2"));
        columnRefMap.put(new ColumnRefOperator(3, Type.BIGINT, "v3", true),
                t0.getBaseColumn("v3"));

        OptExpression opt =
                new OptExpression(new LogicalOlapScanOperator(t0, columnRefMap, Maps.newHashMap(), null, -1, null));
        Assert.assertTrue(Utils.hasUnknownColumnsStats(opt));

        GlobalStateMgr.getCurrentStatisticStorage().addColumnStatistic(t0, "v2",
                new ColumnStatistic(1, 1, 0, 1, 1));
        GlobalStateMgr.getCurrentStatisticStorage().addColumnStatistic(t0, "v3",
                new ColumnStatistic(1, 1, 0, 1, 1));
        opt = new OptExpression(new LogicalOlapScanOperator(t0, columnRefMap, Maps.newHashMap(), null, -1, null));
        Assert.assertFalse(Utils.hasUnknownColumnsStats(opt));
    }

    @Test
    public void unknownStats2() {
        GlobalStateMgr globalStateMgr = connectContext.getGlobalStateMgr();
        OlapTable t1 = (OlapTable) globalStateMgr.getDb("test").getTable("t1");
        OptExpression opt =
                new OptExpression(
                        new LogicalOlapScanOperator(t1, Maps.newHashMap(), Maps.newHashMap(), null, -1, null));
        Assert.assertFalse(Utils.hasUnknownColumnsStats(opt));
    }

    @Test
    public void testCountJoinNode() {
        OptExpression root = OptExpression.create(
                new LogicalJoinOperator(JoinOperator.LEFT_OUTER_JOIN, null),
                OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_OUTER_JOIN, null),
                        OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null)),
                        OptExpression.create(new LogicalValuesOperator(Lists.newArrayList(), Lists.newArrayList()))),
                OptExpression.create(new LogicalValuesOperator(Lists.newArrayList(), Lists.newArrayList())));

        assertEquals(1, Utils.countJoinNodeSize(root, JoinOperator.semiAntiJoinSet()));


        //      outer join (left child semi join node = 1, right child semi join node = 3) => result is 3
        //      /         \
        //  semi join      semi join (left child semi join node = 1, right child semi join node = 1)
        //                   /    \            => result is 1 + 1 + 1 = 3
        //           outer join   semi join
        //             /      \
        //         semi join   node
        root = OptExpression.create(
                new LogicalJoinOperator(JoinOperator.LEFT_OUTER_JOIN, null),
                OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null)),
                OptExpression.create(new LogicalProjectOperator(Maps.newHashMap()),
                        OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null),
                                OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_OUTER_JOIN, null),
                                        OptExpression.create(
                                                new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null)),
                                        OptExpression.create(
                                                new LogicalValuesOperator(Lists.newArrayList(), Lists.newArrayList()))),
                                OptExpression.create(
                                        new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null)))));

        assertEquals(3, Utils.countJoinNodeSize(root, JoinOperator.semiAntiJoinSet()));

        //      semi join (left child semi join node = 0, right child semi join node = 3) => result is 0 + 3 + 1 = 4
        //      /         \
        //  inner join   semi join (left child semi join node = 2, right child semi join node = 0)
        //                /    \            => result is 2 + 0 + 1 = 3
        //        semi join    node
        //         /      \
        //   semi join  node
        root = OptExpression.create(
                new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null),
                OptExpression.create(new LogicalJoinOperator(JoinOperator.INNER_JOIN, null)),
                OptExpression.create(new LogicalProjectOperator(Maps.newHashMap()),
                        OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null),
                                OptExpression.create(new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null),
                                        OptExpression.create(
                                                new LogicalJoinOperator(JoinOperator.LEFT_SEMI_JOIN, null)),
                                        OptExpression.create(
                                                new LogicalValuesOperator(Lists.newArrayList(), Lists.newArrayList()))),
                                OptExpression.create(
                                        new LogicalValuesOperator(Lists.newArrayList(), Lists.newArrayList())))));

        assertEquals(4, Utils.countJoinNodeSize(root, JoinOperator.semiAntiJoinSet()));
    }
}

File: be/test/storage/utils_test.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "storage/utils.h"

#include <gtest/gtest.h>

#include "runtime/mem_tracker.h"

namespace starrocks {
class TestUtils : public ::testing::Test {};
TEST_F(TestUtils, test_valid_decimal) {
    ASSERT_TRUE(valid_decimal("0", 0, 0));
    ASSERT_TRUE(valid_decimal("0", 2, 2));
    ASSERT_TRUE(valid_decimal("0", 10, 10));
    ASSERT_TRUE(valid_decimal("0.618", 10, 10));
    ASSERT_TRUE(valid_decimal("0.0618", 10, 10));
    ASSERT_TRUE(valid_decimal("0.0", 10, 10));
    ASSERT_TRUE(valid_decimal("-0.618", 10, 10));
    ASSERT_TRUE(valid_decimal("-0.0618", 10, 10));
    ASSERT_TRUE(valid_decimal("-0.0", 10, 10));
    ASSERT_FALSE(valid_decimal("3.14", 10, 10));
    ASSERT_FALSE(valid_decimal("31.4", 10, 10));
    ASSERT_FALSE(valid_decimal("314.15925", 10, 10));
    ASSERT_FALSE(valid_decimal("-3.14", 10, 10));
    ASSERT_FALSE(valid_decimal("-31.4", 10, 10));
    ASSERT_FALSE(valid_decimal("-314.15925", 10, 10));
    ASSERT_TRUE(valid_decimal("-3.14", 3, 2));
    ASSERT_TRUE(valid_decimal("-31.4", 3, 1));
    ASSERT_TRUE(valid_decimal("-314.15925", 8, 5));
    ASSERT_TRUE(valid_decimal("3.14", 3, 2));
    ASSERT_TRUE(valid_decimal("31.4", 3, 1));
    ASSERT_TRUE(valid_decimal("314.15925", 8, 5));
}

TEST_F(TestUtils, test_is_tracker_hit_hard_limit) {
    std::unique_ptr<MemTracker> tracker = std::make_unique<MemTracker>(1000, "test", nullptr);
    tracker->consume(2000);
    ASSERT_TRUE(is_tracker_hit_hard_limit(tracker.get(), 0.1));
    ASSERT_TRUE(is_tracker_hit_hard_limit(tracker.get(), 1.1));
    ASSERT_TRUE(is_tracker_hit_hard_limit(tracker.get(), 1.5));
    ASSERT_TRUE(is_tracker_hit_hard_limit(tracker.get(), 1.7));
    ASSERT_TRUE(!is_tracker_hit_hard_limit(tracker.get(), 2));
    ASSERT_TRUE(!is_tracker_hit_hard_limit(tracker.get(), 2.5));
    ASSERT_TRUE(!is_tracker_hit_hard_limit(tracker.get(), 3));
    ASSERT_TRUE(!is_tracker_hit_hard_limit(tracker.get(), 4));
}

TEST_F(TestUtils, test_valid_datetime) {
    ASSERT_TRUE(valid_datetime("2020-01-01 00:00:00"));
    ASSERT_TRUE(valid_datetime("2020-01-01 00:00:00.0"));
    ASSERT_TRUE(valid_datetime("2020-01-01 00:00:00.01"));
    ASSERT_TRUE(valid_datetime("2020-01-01 00:00:00.012345"));
    ASSERT_TRUE(valid_datetime("2020-01-01 00:00:00.1230"));
    ASSERT_FALSE(valid_datetime("2020-01-01 00:00:00.0123456"));
}

TEST_F(TestUtils, test_parse_data_size) {
    // Empty string should return 0
    ASSERT_EQ(0, parse_data_size(""));

    // String with only spaces should return 0
    ASSERT_EQ(0, parse_data_size("   "));

    // String with invalid unit should return 0
    ASSERT_EQ(0, parse_data_size("10Bytes"));
    ASSERT_EQ(0, parse_data_size("10X"));
    ASSERT_EQ(0, parse_data_size("abc"));
    ASSERT_EQ(0, parse_data_size("10.5XB"));

    // Only number, no unit, should be treated as bytes
    ASSERT_EQ(10, parse_data_size("10"));
    ASSERT_EQ(123, parse_data_size("123"));

    // Test with 'B' and spaces
    ASSERT_EQ(10, parse_data_size("10B"));
    ASSERT_EQ(10, parse_data_size("  10B  "));
    ASSERT_EQ(10, parse_data_size("10 b"));
    ASSERT_EQ(10, parse_data_size("10 b "));

    // Test with K/k/KB/kb
    ASSERT_EQ(10 * 1024, parse_data_size("10K"));
    ASSERT_EQ(10 * 1024, parse_data_size("10KB"));
    ASSERT_EQ(10 * 1024, parse_data_size("10k"));
    ASSERT_EQ(10 * 1024, parse_data_size("10kb"));
    ASSERT_EQ(10 * 1024, parse_data_size(" 10 KB "));

    // Test with M/m/MB/mb
    ASSERT_EQ(10 * 1024 * 1024, parse_data_size("10M"));
    ASSERT_EQ(10 * 1024 * 1024, parse_data_size("10MB"));
    ASSERT_EQ(10 * 1024 * 1024, parse_data_size("10m"));
    ASSERT_EQ(10 * 1024 * 1024, parse_data_size("10mb"));

    // Test with G/g/GB/gb
    ASSERT_EQ(10LL * 1024 * 1024 * 1024, parse_data_size("10G"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024, parse_data_size("10GB"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024, parse_data_size("10g"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024, parse_data_size("10gb"));

    // Test with T/t/TB/tb
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024, parse_data_size("10T"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024, parse_data_size("10TB"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024, parse_data_size("10t"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024, parse_data_size("10tb"));

    // Test with P/p/PB/pb
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024 * 1024, parse_data_size("10P"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024 * 1024, parse_data_size("10PB"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024 * 1024, parse_data_size("10p"));
    ASSERT_EQ(10LL * 1024 * 1024 * 1024 * 1024 * 1024, parse_data_size("10pb"));

    // Test with decimal numbers
    ASSERT_EQ(1536, parse_data_size("1.5K"));
    ASSERT_EQ(1536, parse_data_size("1.5KB"));
    ASSERT_EQ(1572864, parse_data_size("1.5M"));
    ASSERT_EQ(1610612736, parse_data_size("1.5G"));
    ASSERT_EQ(0, parse_data_size("1.5X")); // invalid unit

    // Test with leading/trailing spaces
    ASSERT_EQ(10 * 1024, parse_data_size("   10K"));
    ASSERT_EQ(10 * 1024, parse_data_size("10K   "));
    ASSERT_EQ(10 * 1024, parse_data_size("   10K   "));

    // Test with spaces between number and unit
    ASSERT_EQ(10 * 1024, parse_data_size("10 K"));
    ASSERT_EQ(10 * 1024, parse_data_size("10   K"));
    ASSERT_EQ(10 * 1024, parse_data_size("  10   K  "));

    // Test with negative number (should return 0)
    ASSERT_EQ(0, parse_data_size("-10K"));

    // Test with value exceeding int64_t max (should return 0)
    ASSERT_EQ(0, parse_data_size("100000000000000000000P"));

    // Test with value below int64_t min (should return 0)
    ASSERT_EQ(0, parse_data_size("-100000000000000000000P"));
}

} // namespace starrocks
